docs/howto/features/identity-provider.md

```diff
@@ -0,0 +1,115 @@
+# Using JupyterHub as an identity provider
+
+JupyterHub can be used as an OAuth2 based identity provider for a community
+managed service of some kind. As an example, a community could themselves host a
+Discourse forum somewhere and login via the community's JupyterHub (that in turn
+may ask users to login by another identity provider).
+
+To help a community use JupyterHub as an identity provider, register an OAuth2
+application with JupyterHub. This is done by declaring an [externally managed
+JupyterHub service] and configuring at least `oauth_client_id` and
+`oauth_redirect_uri`.
+
+[externally managed jupyterhub service]: https://jupyterhub.readthedocs.io/en/stable/reference/services.html#externally-managed-services
+
+## 1. Request required information
+
+We need to configure [`oauth_redirect_uri`], and that is only the community
+managing the service using JupyterHub as an identity provider.
+
+## 2. Configure JupyterHub
+
+1. Define the service in the appropriate hub's `<hub-name>.values.yaml` file:
+
+   ```yaml
+   jupyterhub:
+     hub:
+       services:
+         # name the service to make sense for an URL like
+         # https://<community>.2i2c.cloud/services/<name>
+         <name-of-service>:
+           # The client_id for this service - it *must* start with service-
+           oauth_client_id: service-<name-of-service>
+           # The OAuth2 redirect URI provided to us by the community, such as
+           # https://forum.community.org/auth/oauth2_basic/callback
+           oauth_redirect_uri: <oauth-redirect-uri>
+           # oauth_no_confirm is set to true to avoid a 'do you want to approve
+           # this external service?' authorization prompt. Since this service is
+           # only granted permission to read basic information about the user
+           # and trusted by the community, it shouldn't be relevant
+           oauth_no_confirm: true
+       config:
+         Authenticator:
+           # This prevents the hub home page from showing when the end user
+           # tries to authenticate with the hub from the external service.
+           # Since the hub might send the user to CILogon or GitHub or another
+           # auth service, removing this step makes the flow smoother.
+           auto_login_oauth2_authorize: true
+   ```
+
+2. Configure a `client_secret`
+
+   The JupyterHub service's `api_token` is used as a `client_secret`, lets
+   configure it explicitly.
+
+   First generate an appropriate OAuth2 client secret on the commandline by
+   running `openssl rand -hex 32`, then define appropriate secrets in the hub's
+   `enc-<hub-name>.secret.values.yaml` file:
+
+   ```yaml
+   jupyterhub:
+     hub:
+       services:
+         <name-of-service>:
+           # This will be the client secret
+           api_token: <output-from-openssl>
+   ```
+
+   ````{note}
+   If `api_token` isn't configured explicitly, it gets generated by the
```

Comment:
It was an intentional choice to explicitly set `api_token` for this specific case. So if we had to recreate the cluster or redeploy the hub, it doesn't change! If it does, their identity provider will immediately stop working, and we will need to communicate and co-ordinate this. So that was the reason for explicitly setting it, although right to replicate is also useful.

Reply:
Thanks @yuvipanda, ensuring we can do fresh re-deploys without ensuring retention of this state seems like a more direct motivation worth focusing on!

I removed the details on how to read the k8s secret it and updated the motivation for explicit configuration to mention the need to recoordinate things.

While omitted from this text as it felt a bit out of scope, if the community wants to reduce the complexity of their configuration in the future to not manage such secret, they could opt to not set the `api_token` explicitly at any point in time, the same `api_token` will be still be used.